homeassistant/components/tplink/config_flow.py

```diff
@@ -40,27 +69,114 @@ async def async_step_integration_discovery(
+            self.hass.config_entries.async_reload(entry.entry_id),
+            f"config entry reload {entry.title} {entry.domain} {entry.entry_id}",
         )
+        raise AbortFlow("already_configured")
 
-    async def _async_handle_discovery(self, host: str, mac: str) -> FlowResult:
+    async def _async_handle_discovery(
+        self, host: str, formatted_mac: str, config: Optional[dict] = None
+    ) -> FlowResult:
         """Handle any discovery."""
-        await self.async_set_unique_id(dr.format_mac(mac))
+        current_entry = await self.async_set_unique_id(
+            formatted_mac, raise_on_progress=False
+        )
+        if config and current_entry:
+            self._update_config_if_entry_in_setup_retry(current_entry, host, config)
         self._abort_if_unique_id_configured(updates={CONF_HOST: host})
         self._async_abort_entries_match({CONF_HOST: host})
         self.context[CONF_HOST] = host
         for progress in self._async_in_progress():
             if progress.get("context", {}).get(CONF_HOST) == host:
                 return self.async_abort(reason="already_in_progress")
-
+        credentials = await get_credentials(self.hass)
         try:
-            self._discovered_device = await self._async_try_connect(
-                host, raise_on_progress=True
+            await self._async_try_discover_and_update(
+                host, credentials, raise_on_progress=True
             )
+        except AuthenticationException:
+            return await self.async_step_discovery_auth_confirm()
         except SmartDeviceException:
             return self.async_abort(reason="cannot_connect")
+
         return await self.async_step_discovery_confirm()
 
+    async def async_step_discovery_auth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that auth is required."""
+        assert self._discovered_device is not None
+        errors = {}
+
+        credentials = await get_credentials(self.hass)
+        if credentials and credentials != self._discovered_device.config.credentials:
+            try:
+                device = await self._async_try_connect(
+                    self._discovered_device, credentials
+                )
+            except AuthenticationException:
+                pass  # Authentication exceptions should continue to the rest of the step
+            else:
+                self._discovered_device = device
+                return await self.async_step_discovery_confirm()
+
+        if user_input:
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+            credentials = Credentials(username, password)
+            try:
+                device = await self._async_try_connect(
+                    self._discovered_device, credentials
+                )
+            except AuthenticationException:
+                errors[CONF_PASSWORD] = "invalid_auth"
```

Comment:
I'm not sure if re-using a const meant for a completely different use case is wanted, but it's rather confusing for someone who doesn't know what's going on...

Reply:
This is the standard approach. The error value needs to match the schema key